crates/sui-core/src/transaction_manager.rs

```diff
@@ -107,33 +112,62 @@ impl TransactionManager {
     }
 
-    /// Enqueues certificates into TransactionManager. Once all of the input objects are available
-    /// locally for a certificate, the certified transaction will be sent to execution driver.
-    ///
-    /// REQUIRED: Shared object locks must be taken before calling this function on shared object
-    /// transactions!
-    ///
-    /// TODO: it may be less error prone to take shared object locks inside this function, or
-    /// require shared object lock versions get passed in as input. But this function should not
-    /// have many callsites. Investigate the alternatives here.
     pub(crate) fn enqueue(
         &self,
         certs: Vec<VerifiedExecutableTransaction>,
         epoch_store: &AuthorityPerEpochStore,
     ) -> SuiResult<()> {
-        let inner = &mut self.inner.write();
+        let mut pending = Vec::new();
         for cert in certs {
-            if inner.epoch != cert.epoch() {
+            let digest = *cert.digest();
+            // skip already executed txes
+            if self.authority_store.is_tx_already_executed(&digest)? {
+                // also ensure the transaction will not be retried after restart.
+                let _ = epoch_store.remove_pending_execution(&digest);
+                self.metrics
+                    .transaction_manager_num_enqueued_certificates
+                    .with_label_values(&["already_executed"])
+                    .inc();
+                continue;
+            }
+            let input_object_kinds = cert.data().intent_message.value.input_objects()?;
+            let input_object_keys = self.authority_store.get_input_object_keys(
+                &digest,
+                &input_object_kinds,
+                epoch_store,
+            );
+            pending.push(PendingCertificate {
+                certificate: cert,
+                missing: input_object_keys
+                    .into_iter()
+                    .filter(|key| {
+                        !self
+                            .authority_store
+                            .object_version_exists(&key.0, key.1)
+                            .expect("Checking object existence cannot fail!")
+                    })
+                    .collect(),
+            });
+        }
+
+        // After this point, the function cannot return early and must run to the end. Otherwise,
+        // it can lead to data inconsistencies and potentially some transactions will never get
+        // executed.
+
+        let mut missing_input_objects = Vec::new();
+        // Lock is held only for updating internal state.
+        let mut inner = self.inner.write();
+
+        for pending_cert in pending {
+            if inner.epoch != pending_cert.certificate.epoch() {
                 warn!(
                     "Ignoring enqueued certificate from wrong epoch. Expected={} Certificate={:?}",
-                    inner.epoch, cert
+                    inner.epoch, pending_cert.certificate
```

Comment:
should we also remove from the pending table?

Reply:
Not sure if this is possible but zero harm in doing this.